apps/lib/app_params.c

```diff
@@ -94,6 +97,56 @@ int print_param_types(const char *thing, const OSSL_PARAM *pdefs, int indent)
     return 1;
 }
 
+/* Output the body of a UTF8 string which might not be zero terminated */
+static void print_param_utf8(const char **s_ptr, size_t len)
+{
+    const char *s;
+
+    if (s_ptr == NULL) {
+        BIO_puts(bio_out, " ptr null\n");
+        return;
+    }
+    if ((s = *s_ptr) == NULL) {
+        BIO_puts(bio_out, " null\n");
+        return;
+    }
+    BIO_puts(bio_out, "'");
+    if (len > 0)
+        BIO_write(bio_out, s, len);
```

Comment:
If the string is really supposed to be UTF8, should this be checked here?
Should anything be done about control characters?

Reply:
Well, the current behaviour may not be ideal, but I guess these are RNG provider parameters (the function is only used for these, and could if that's unlikely to change be renamed), and might therefore never contain particularly exotic strings.

As for how to check for UTF8, loop over the bytes reading one UTF8 code point at a time, until your reach the end or find an invalid one: https://github.com/openssl/openssl/blob/28de1f5004c1083d358e6934552124a201e0251e/crypto/asn1/a_mbstr.c#L229-L233
One might also, as/when appropriate, reject internal NUL bytes, and in some cases accept non-minimal Java-style `0xC080` as a non-NUL placeholder for semantic NUL bytes in UTF8 C-strings.

Reply:
I still don't see the point: the parameter's value is coming from a trusted source.

Reply:
Agreed

Reply:
UTF8 cannot have embedded NUL characters. Any multi-byte UTF8 code is indicated with the 0x80 bit on.

Reply:
Actually, UTF8 absolutely can have embedded NUL bytes, the use of 0xC080 is language-ecosystem-specific hack in Java, and subsequently a few others. This is actually a mild violation of UTF8 rules and some validators will reject such an encoding as malformed.

Actual UTF8 (e.g. ASCII which is a subset) with embedded NUL bytes was notably observed when Moxie managed to obtain a certificate for `*<NUL>.<some>.<domain>` and various things had to be urgently patched (in Postfix I was already checking for internal NULs for a few years, and nothing needed to be done).